### PR TITLE
Cleaned up HeatSourceBar and Added ADMatHeatSource

### PR DESCRIPTION
**Pull request by tophmatthews**
This adds `ADMatHeatSource` which is essentially a body force given a scalar and a material property. In the process, I cleaned up the test for the `HeatSource` heat_source_bar.i. It was originally in the `combined` app since it used `Density` which is in `misc`, but there are not displacements so density can just be a `GenericConstatMaterial`. I moved `heat_source_bar` into the `HeatConduction` module, removed the 2D custom mesh and journal file in favor of a 1D generated mesh, and simplified the input file greatly. I changed the gold file only due to the reduction in mesh dimension, but there was no change in the answer. I added a `NodalL2Error` postprocessor to compare to the analytical solution as well. `ad_heat_source_bar.i` is the same thing as `heat_source_bar`, but with AD kernels only.

(#12633)

### CODE REVIEW

**Comment by lindsayad** [comment]
Getting an exodiff, number of time steps differ?

**Comment by tophmatthews** [comment]
Yeah, I'm not too sure why that is the case...

File: modules/heat_conduction/include/kernels/ADMatHeatSource.h [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+class ADMatHeatSource : public ADKernel<compute_stage>
+{
+public:
+  ADMatHeatSource(const InputParameters & parameters);
+
+protected:
+  virtual ADResidual computeQpResidual();
+
+  const Real _scalar;
+  const ADMaterialProperty(Real) & _material;
```
Comment (lindsayad) [inline modules/heat_conduction/include/kernels/ADMatHeatSource.h]: Perhaps `_material_property` or `_mat_prop` instead?

File: modules/heat_conduction/src/kernels/ADMatHeatSource.C [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+#include "ADMatHeatSource.h"
+
+registerADMooseObject("HeatConductionApp", ADMatHeatSource);
+
+defineADValidParams(
+    ADMatHeatSource,
+    ADKernel,
+    params.addParam<Real>("scalar", 1.0, "Scalar multiplied by the body force term");
+    params.addRequiredParam<MaterialPropertyName>("material",
+                                                  "Material multiplied by the body force term"););
```
Comment (lindsayad) [inline modules/heat_conduction/src/kernels/ADMatHeatSource.C]: Modeling after the `BodyForce` kernel, perhaps this doc string should be "Material property describing the body force". If both `scalar` and the `_mat_prop` are multiplied by the body force term, then what is the body force term?

File: modules/heat_conduction/src/kernels/ADMatHeatSource.C [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+#include "ADMatHeatSource.h"
+
+registerADMooseObject("HeatConductionApp", ADMatHeatSource);
+
+defineADValidParams(
+    ADMatHeatSource,
+    ADKernel,
+    params.addParam<Real>("scalar", 1.0, "Scalar multiplied by the body force term");
+    params.addRequiredParam<MaterialPropertyName>("material",
```
Comment (lindsayad) [inline modules/heat_conduction/src/kernels/ADMatHeatSource.C]: I would rename this to whatever you rename the kernel member (minus the leading underscore)

File: modules/heat_conduction/test/tests/heat_source_bar/ad_heat_source_bar.i [hunk 4]
```diff
@@ -0,0 +1,89 @@
+#This is a simple 1D test of the volumetric heat source with material properties
```
Comment (lindsayad) [inline modules/heat_conduction/test/tests/heat_source_bar/ad_heat_source_bar.i]: Not critical, but suggest space after the `#`

**Review by lindsayad — CHANGES_REQUESTED** [review]
Looks good. Just a couple of cosmetic requests

File: modules/heat_conduction/src/kernels/ADMatHeatSource.C (hunk 2 above)
Comment (tophmatthews) [reply modules/heat_conduction/src/kernels/ADMatHeatSource.C]: I modeled this after the constant coefficient in `BodyForce`
```
params.addParam<Real>("value", 1.0, "Coefficient to multiply by the body force term");
```
I agree it sounds a bit weird. I'll change it to what you suggested

**Comment by lindsayad** [comment]
Yes, but the material is kind of an even more generic stand-in (like we
discussed) for the "function" param in BodyForce whose doc string is "A
function that describes the body force"

**Comment by tophmatthews** [comment]
@lindsayad ready for you again to check

**Review by lindsayad — APPROVED** [review]

**Comment by tophmatthews** [comment]
@lindsayad This added something real screwy https://civet.inl.gov/event/37495/ but it seems to deal with ADThermoDiffusion?!

**Comment by tophmatthews** [comment]
ADThermoDiffusion should be `registerADMooseObject("MiscApp", ADThermoDiffusion);`not `MooseApp`, but that doesn't prevent the error.

**Comment by tophmatthews** [comment]
Oh, crap, it's the header guards...

**Comment by lindsayad** [comment]
@tophmatthews It looks like we're getting some exodiffs on [HPC testing](https://civet.inl.gov/job/271573/). Looks like you might just want to raise the floor for those tests (I think maybe the `abs_zero` in the `tests` spec file is one option...default is 1e-10, perhaps a change to 5e-8 or something like that which would be more in line with our default non-linear tolerance being used here)

**Comment by tophmatthews** [comment]
@lindsayad this is failing on the `NodalL2Error` postprocessor
```
error  rel diff:  3.4412847e-09 ~  1.2921338e-08 = 7.33674e-01 (FAILED)
```
I'm not sure how you guys handle those small changes in the error. It may be worth not even outputting it to exodus rather that really loosening the tolerances?

**Comment by tophmatthews** [comment]
I see, the abs_zero is typically done

**Pull request merged**